R/estimate_risk.R

```diff
@@ -213,393 +220,495 @@ estimate_risk <- function(
+    if(fit5$converged == FALSE)
+      fit5$risks_start = "_start"
+
+    fit6 <- possibly_estimate_margstd_boot(
+      formula = formula,
+      data = data,
+      estimand = estimand,
+      variable = variable,
+      at = at,
+      ...
+    )
+
+    fit7 <- possibly_estimate_margstd_delta(
+      formula = formula,
+      data = data,
+      estimand = estimand,
+      variable = variable,
+      at = at,
+      ...
+    )
+
+    # If RR requested, add on case-duplication model and, for comparison,
+    # the plain logistic model
+    if(estimand == "rr") {
+
+      fit8 <- possibly_estimate_logistic(
         formula = formula,
         data = data,
-        estimand = estimand,
-        variable = variable,
-        at = at,
-        ...)
+        ...
+      )
 
-      fit7 <- possibly_estimate_margstd_delta(
+      fit9 <- possibly_estimate_duplicate(
         formula = formula,
         data = data,
-        estimand = estimand,
-        variable = variable,
-        at = at,
-        ...)
-
-      # If RR requested, add on case-duplication model and, for comparison,
-      # the plain logistic model
-      if(estimand == "rr") {
-        fit8 <- possibly_estimate_logistic(
-          formula = formula,
-          data = data,
-          ...)
-        fit9 <- possibly_estimate_duplicate(
+        ...
+      )
+
+      if(!is.null(coef(fit9)))  # attempt only if 'duplicate' converged
+        fit10 <- possibly_estimate_glm_startd(
           formula = formula,
           data = data,
-          ...)
-
-        if(!is.null(coef(fit9)))  # attempt only if 'duplicate' converged
-          fit10 <- possibly_estimate_glm_startd(
-            formula = formula,
-            data = data,
-            link = link,
-            start = coef(fit9),
-            start_type = "d",
-            ...)
-        else  # make possibly_estimate_glm return a non-converged object
-          fit10 <- possibly_estimate_glm_startd(
-            formula = "nonsense",
-            data = "nodata",
-            start_type = "d")
-
-        fit1$all_models = list(
-          robpoisson = fit1,
-          glm = fit2,
-          glm_startp = fit3,
-          glm_cem = fit4,
-          glm_cem_startp = fit5,
-          margstd_boot = fit6,
-          margstd_delta = fit7,
-          logistic = fit8,
-          duplicate = fit9,
-          glm_startd = fit10)
-      } else
-        fit1$all_models = list(
-          robpoisson = fit1,
-          glm = fit2,
-          glm_start = fit3,
-          glm_cem = fit4,
-          glm_cem_startp = fit5,
-          margstd_boot = fit6,
-          margstd_delta = fit7)
-      fit1
-    },
-
-    # Specific models that were directly requested
-    robpoisson = estimate_poisson(
+          link = link,
+          start = coef(fit9),
+          start_type = "d",
+          ...
+        )
+      #TODO seeing a pattern: let's make returning a non-converged object
+      # an internal function that can be used in all cases like this
```

Comment:
Right - such a function just exists under different names because it returns slightly different object types depending on what model was supposed to be fit.
https://github.com/stopsack/risks/blob/f48f0a05ce5de3a6a729691847e2cdcd8f24554f/R/zestimate_risk-utils.R#L68-L117